### PR TITLE
feat: removing cx and cy, testing collapse node and lifting, some major refactor

### PR DESCRIPTION
**Pull request by christojeffrey**

### CODE REVIEW

File: svelte/src/helper/index.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import type { EdgeType, GraphDataEdge, SimpleNode } from '$types';
 
 // debugging purpose
-export function debuggingConsole(...args: never[]) {
+export function debuggingConsole(...args: unknown[]) {
```
Comment (christojeffrey) [inline svelte/src/helper/index.ts:4]: this is done to fix linter

File: svelte/src/routes/+page.svelte [hunk 2]
```diff
@@ -37,11 +34,13 @@
...
 		nodeCornerRadius: 5,
 		nodePadding: 5,
+		textSize: 10,
 		shownEdgesType: new Map<EdgeType, boolean>(),
 		showEdgeLabels: false,
 		showNodeLabels: true,
 		nodeDefaultColor: '#6a6ade',
-		nodeColors: ['#32a875', '#d46868']
+		nodeColors: ['#32a875', '#d46868'],
+		disableAnimation: false
```
Comment (christojeffrey) [inline svelte/src/routes/+page.svelte:43]: add new disable animation feature. so that on every filter update, the node is not dragged to the middle of the screen again

File: svelte/src/routes/+page.svelte [hunk 3]
```diff
@@ -52,16 +51,34 @@
 
 	let isMounted = false;
 
-	function handleNodeCollapseClick(datum: GraphDataNode) {
-		onNodeCollapseClick(datum, config, () => {
-			doRefilter = true;
-		});
+	function handleNodeCollapseClick(clickedNode: GraphDataNode) {
```
Comment (christojeffrey) [inline svelte/src/routes/+page.svelte:54]: move the handler inside here. previously, filter can be done in two part. on click, and on filter function. now to make it more consistent, filter only gonna be done inside the filter function.

Comment (christojeffrey) [reply svelte/src/routes/+page.svelte:54]: previously we need to handle 'doUncollapse', that's why the handler is created inside filter file. how come we don't need to do that now? 

current implementation: the action of filtering is going to be reset on each filter. then the filter is gonna collapse or lift based on config. 

other options: do the collapse and uncollapse directly after click. this will alleviate the need for resetting on every filter, but this will lead to multiple redirection of edge. I think to handle this, we need to note down the trace of edges redirection. which is ... complicated. might introduce more confusion in the future.

basically it's a trade between code complexity and efficiency. I usually go for code complexity. But I think the status quo is complicated enough, and code complexity seems to be our bottleneck to go forward instead of the performance of the application. So I took worse efficiency, i.e. resetting on each filter

File: svelte/src/types/index.ts [hunk 4]
```diff
@@ -84,27 +86,28 @@ export interface GraphDataEdge extends d3.SimulationLinkDatum<GraphDataNode> {
...
-	members?: GraphDataNode[];
+	members: GraphDataNode[];
 	parent?: GraphDataNode;
 
-	outgoingLinks?: GraphDataEdge[];
-	incomingLinks?: GraphDataEdge[];
+	outgoingLinks: GraphDataEdge[];
+	incomingLinks: GraphDataEdge[];
+
+	originalOutgoingLinks: GraphDataEdge[];
```
Comment (christojeffrey) [inline svelte/src/types/index.ts:103]: this part is going to help on filter reset. so that we don't have to traverse each link on each filter, reducing the computation needed. trade between speed and memory. I think memory is not a bottleneck right now (I guess apps right now just doesn't care about memory usage in general too lol).

File: svelte/src/scripts/create-graph-data/index.ts [hunk 5]
```diff
@@ -16,27 +16,35 @@ function assignParentReference(nodes: GraphDataNode[]) {
 				arr[index].parent = node;
 			});
 			assignParentReference(node.members);
+			node.originalMembers = node.members;
 		}
 	});
 }
 
-function assignLinkReference(
+function assignOutgoingAndIncomingLinksAndOriginalSourceAndTargetReference(
```
Comment (christojeffrey) [inline svelte/src/scripts/create-graph-data/index.ts:24]: previously the the edges and nodes attribute can be undefined (has question mark in it). specifically: members, incomingLinks, outgoingLinks. but now they're a must. if there's none, use empty array []. this makes the handling much more consistent and shorter, no need to wrap them like below
```javascript
if(someVariable){
  someVariable.action(); // can't just use question mark. linter seems to go mad
}

someArray || [].length > 0
```
this function is used to inject them.

File: svelte/src/scripts/draw/helper/drag-handler.ts [hunk 6]
```diff
@@ -9,8 +9,8 @@ export function dragStartedNode(
 	simulations.forEach((simulation) => {
 		if (!event.active) simulation.alpha(slowAlpha).restart();
 	});
-	event.subject.fx = event.subject.x;
-	event.subject.fy = event.subject.y;
+	event.subject.fx = null;
```
Comment (christojeffrey) [inline svelte/src/scripts/draw/helper/drag-handler.ts:12]: this is one of the changes necessary to remove calculatedx and calculated y (cx and cy). 
if drag-handler.ts is left the same, it will jitter. 

my guess: I think it's because there's 2 part that can change the x and y. this guy, and our function inside tick. the x and y is overwritten back and forth, by fx (I think it's **forced x** ) and fy; and our function; resulting in jitter. 

why it was done this way? why do we use fx and fy at first? because that's the example I got (either from the documentation, or from stackoverflow or forums like it)

File: svelte/src/scripts/draw/helper/element-adder.ts [hunk 7]
```diff
@@ -36,15 +36,16 @@ export function addNodeContainerElements(
...
+	drawSettings: DrawSettingsInterface
 ) {
 	return nodeContainer
 		.append('text')
 		.attr('class', 'node-label')
 		.attr('text-anchor', 'middle')
 		.attr('dominant-baseline', 'middle')
 		.attr('fill', 'black')
-		.attr('font-size', '10px')
+		.attr('font-size', drawSettings.textSize + 'px')
```
Comment (christojeffrey) [inline svelte/src/scripts/draw/helper/element-adder.ts:48]: add textSize so that the top padding of the text inside node is based on this

File: svelte/src/scripts/draw/helper/tick/node-ticked.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	containerElement: d3.Selection<SVGGElement, GraphDataNode, SVGGElement, unknown>,
+	nodeElements: d3.Selection<SVGRectElement, GraphDataNode, SVGGElement, unknown>,
+	drawSettings: DrawSettingsInterface,
+	nodeLabelsElements: d3.Selection<SVGTextElement, GraphDataNode, SVGGElement, unknown>,
+	collapseButtonElements: d3.Selection<SVGCircleElement, GraphDataNode, SVGGElement, unknown>,
+	liftButtonElements: d3.Selection<SVGCircleElement, GraphDataNode, SVGGElement, unknown>
+) {
+	// calculate nodes width and height, x and y (which is cx and cy) - based on the location of the its members.
+	// only do this calculation once, on master simulation
+	calculateWidthHeighXandYBasedOnChildrenRecursive(graphData.nodes, drawSettings);
```
Comment (christojeffrey) [inline svelte/src/scripts/draw/helper/tick/node-ticked.ts:33]: now you can see clearly the difference between master simulation and inner simulation

File: svelte/src/scripts/filter/collapse-nodes.ts [hunk 9]
```diff
@@ -1,88 +1,13 @@
...
-			if (flatChildrenId.includes(link.target.id)) {
-				link.originalTarget = link.target;
-				link.target = graphData.flattenNodes[collapsedNodeIndex];
-				const key = `${link.source.id}-${link.target.id}`;
-				duplicateLinks.set(key, [...(duplicateLinks.get(key) ?? []), link]);
-			}
-		});
+		// redirect link
+		const redirectDestination = collapsedNode;
+		redirectAllEdgeToDestinationNode(redirectDestination, collapsedNode);
```
Comment (christojeffrey) [inline svelte/src/scripts/filter/collapse-nodes.ts:8]: collapse node and lift edges is very much related. 

collapse node is the superset of lift edges. so let's make use the same function

File: svelte/src/scripts/filter/index.ts [hunk 10]
```diff
@@ -1,11 +1,37 @@
...
-	graphData.links = liftDependencies(config, graphData);
-
+	liftDependencies(config);
 	// handle collapsed vertices
-	doCollapseNodes(config, graphData);
+	doCollapseNodes(config);
+}
+
+function doCombineEdgesWeight(config: ConfigInterface, graphData: GraphData) {
+	// TODO: impement this, and move it to another file
```
Comment (christojeffrey) [inline svelte/src/scripts/filter/index.ts:18]: combining edges weight was  was done inside lift edges and collapse node. I think we can move them here, in one place. not yet implemented. but should use the same function as previous. just pushing this here since I don't think it's gonna be blocking, and can be done parallel with other things, which means it can be done later.

File: vanilla/index.html
Comment (christojeffrey) [inline vanilla/index.html:1]: remove unused folder

File: svelte/src/scripts/converter/index.ts [hunk 11]
```diff
@@ -59,6 +59,7 @@ export function converter(rawData: RawInputType): ConvertedData {
 			}
 		});
 	}
+	console.log(nodes);
```
Comment (Adityapnn811) [inline svelte/src/scripts/converter/index.ts]: Might want to remove this

**Comment by Adityapnn811** [comment]
image
I tried running it locally and it bugs out. Does this happen to you too?

**Comment by christojeffrey** [comment]
hem is it like that from the beginning? can you explain the step to reproduce it?

**Comment by christojeffrey** [comment]
wops no need to reply, I am on it already. it's like this after merging with main

**Comment by christojeffrey** [comment]
@Adityapnn811 what if we disable it first, then fix it in another commit so that we can work parallely?

**Review by Adityapnn811 — CHANGES_REQUESTED** [review]

File: svelte/src/scripts/converter/index.ts (hunk 11 above)
Comment (christojeffrey) [reply svelte/src/scripts/converter/index.ts]: alright donee

**Review by Adityapnn811 — APPROVED** [review]
